lib/kubernetes-deploy/kubernetes_resource.rb

```diff
@@ -54,19 +41,28 @@ def timeout
       self.class.timeout
     end
 
-    def initialize(name:, namespace:, context:, file:, logger:)
+    def initialize(namespace:, context:, template:, logger:)
       # subclasses must also set these if they define their own initializer
-      @name = name
+      @name = template.fetch("metadata", {})["name"]
```

Comment:
At least `nil` would fail and we'd know when and where that happened. With an empty string we may never know if it happens. Also IMO `nil` if better from the semantics perspective because NULL value kind of unknown value.